ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -195,14 +268,132 @@ static INT_PTR CALLBACK ShowMessageExtProc(HWND hwndDlg, UINT uMsg, WPARAM wPara
+        (GetSystemMetrics(SM_CYSCREEN) / 2) - ((rect.bottom - rect.top) / 2),
+        rect.right - rect.left, rect.bottom - rect.top, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDOK), 256, 224, 77, 27, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDCANCEL), 340, 224, 77, 27, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc3), 232, 56, 72, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDC_LABEL), 8, 56, 72, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc2), 8, 8, 93, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc4), 232, 176, 50, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc1), 8, 24, 409 * 100, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDC_DNAME), 8, 24, 409, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, lst1), 232, 72, 185, 93, TRUE);
+      MoveWindow(GetDlgItem(hWnd, lst2), 8, 72, 213, 123, TRUE);
+      MoveWindow(GetDlgItem(hWnd, cmb2), 232, 192, 185, 19, TRUE);
+      ShowWindow(GetDlgItem(hWnd, stc1), SW_HIDE);
+      DlgDirListW(hWnd, wstr_dname, lst1, stc1, DDL_ARCHIVE | DDL_READWRITE | DDL_READONLY);
+      GetDlgItemTextW(hWnd, stc1, wstr_stc1, MAX_PATH);
+      static string DlgItemText;
+      DlgItemText = shorten(LowerCaseToActualPathName(wstr_stc1));
+      ActualPath = widen(string_replace_all(DlgItemText + "\\", "\\\\", "\\"));
+      SetDlgItemTextW(hWnd, IDC_DNAME, ActualPath.c_str());
+      PostMessageW(hWnd, WM_SETFOCUS, 0, 0);
+    }
+  break;
+
+  case WM_CREATE:
+    {
+      SetWindowLongPtr(hWnd, GWL_STYLE, GetWindowLongPtr(hWnd, GWL_STYLE) | DS_FIXEDSYS);
+      SetWindowPos(hWnd, 0, 0, 0, 0, 0, SWP_FRAMECHANGED | SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER);
+    }
+
+  case WM_PAINT:
+    {
+      GetDlgItemTextW(hWnd, stc1, wstr_stc1, MAX_PATH);
+      static string DlgItemText;
+      DlgItemText = shorten(LowerCaseToActualPathName(wstr_stc1));
+      ActualPath = widen(string_replace_all(DlgItemText + "\\", "\\\\", "\\"));
+      SetDlgItemTextW(hWnd, IDC_DNAME, ActualPath.c_str());
+    }
+  break;
+
+  case WM_COMMAND:
+    {
+      if (HIWORD(wParam) == LBN_DBLCLK && LOWORD(wParam) == lst1)
+        return TRUE;
+      else if (HIWORD(wParam) == BN_CLICKED && LOWORD(wParam) == IDOK)
+      {
+        wcsncpy(wstr_dname, ActualPath.c_str(), MAX_PATH);
+        PostMessageW(hWnd, WM_COMMAND, IDABORT, 0);
+        return TRUE;
+      }
+      else if (HIWORD(wParam) == BN_CLICKED && LOWORD(wParam) == IDCANCEL)
+      {
+        tstring tstr_dname = widen("");
+        wcsncpy(wstr_dname, tstr_dname.c_str(), MAX_PATH);
+        PostMessageW(hWnd, WM_COMMAND, IDABORT, 0);
+        return TRUE;
+      }
+    }
+  break;
+
+  case WM_CLOSE:
+    {
+      tstring tstr_dname = widen("");
+      wcsncpy(wstr_dname, tstr_dname.c_str(), MAX_PATH);
+      PostMessageW(hWnd, WM_COMMAND, IDABORT, 0);
+      return TRUE;
+    }
+  break;
+  }
+
+  return FALSE;
+}
+
+UINT_PTR CALLBACK CCHookProc(HWND hdlg, UINT uiMsg, WPARAM wParam, LPARAM lParam)
```

Comment:
I used that function name for the hook procedure as a copy and paste from straight out of the Microsoft docs. I can rename it though, that's fine. The hook procedure in this case does 2 useful things. 1) Allows you to change the title bar text like you guys originally gave almost every function the option to do but sadly not this one and 2) I made the position on the screen the window opens less awkward than the default by opening it at the same position GameMaker's get_color opens.

Reply:
I think you are looking at the doc the wrong way. If it's not an actual source code block you can't just copy and paste the function signature because it is ambiguous. C++ will look at it and not know if CCHookProc means yours or one in another translation unit.

I am fine with the caption thing, but where was it opening and where does GMS open it? I don't quite understand, I thought both would just open it centered. Please clarify.